config/tls_termination.go

```diff
@@ -1,10 +1,81 @@
 package config
 
+type TLSTerminationLocation int
+
+const (
+	// Terminate TLS at the ngrok edge. The backend will receive a plaintext
+	// stream.
+	TLSAtEdge TLSTerminationLocation = iota
+	// Terminate TLS in the ngrok library. The library will receive the
+	// handshake and perform TLS termination, and the backend will receive the
+	// plaintext stream.
+	// TODO: export this once implmented
+	tlsAtLibrary
+)
+
+type tlsTermination struct {
+	location TLSTerminationLocation
+	key      []byte
+	cert     []byte
+}
+
+func (tt tlsTermination) ApplyTLS(cfg *tlsOptions) {
+	switch tt.location {
+	case tlsAtLibrary:
+		cfg.KeyPEM = nil
+		cfg.CertPEM = nil
+		// TODO: implement this in the tunnel `Accept` call.
+		panic("automatic tls termination in-app is not yet supported")
```

Comment:
can we return this as an error to the caller instead of panic'ing? if not, we should probably document it loudly

Reply:
Not in an ergonomic way. You'd have to do something like:

```go
opt, err := config.WithTerminateTLSAt(config.TLSAtEdge)
if err != nil {
  return err
}

tun, err := config.TLSEndpoint(opt)
// etc.
```

rather than calling the functional option inline. Could probably rearchitect it such that the options are run at tunnel bind time, and then we could surface any configuration errors in the listen call, but that doesn't feel worth doing just for this.

I'm OK with it panicing for now since the constant isn't exported. Someone would have to cast an integer like `config.WithTLSTerminationAt(config.TLSTerminationLocation(1))`, and in that case I feel like a panic is well-deserved :grin:

Reply:
> Could probably rearchitect it such that the options are run at tunnel bind time, and then we could surface any configuration errors in the listen call, but that doesn't feel worth doing just for this.

we could also just record the error and surface it at listen time? agreed not a high priority